[Flow] Add options for targeted tracing/breaking on dispatches

qedawkins opened the pull request:
Opening a PR to see if this or something similar would be wanted in core IREE. Essentially this enables cropping the input program at a specified dispatch. We have been using this in SHARK-Runtime as a debugging tool.

For example, for certain backends and/or large models, `--iree-flow-trace-dispatch-tensors` is far too slow and produces too much output to easily iterate with so being able to quickly "binary search" can enable much faster isolation of correctness issues. This also has proven useful for isolating "soft" crashes/unwanted behavior like codegen blowing up (happens sometimes in SPIR-V).

The current implementation here is a bit hacky (and assumes single output everywhere) so feedback on a "correct" way to implement this is welcome.

qedawkins committed:
[Flow] Add pass for breaking/tracing on targeted dispatches

We have been using this in SHARK-Runtime as a debugging tool.

For example, for certain backends and/or large models,
`--iree-flow-trace-dispatch-tensors` is far too slow and produces too
much output to easily iterate with so being able to quickly "binary
search" can enable much faster isolation of correctness issues. This
also has proven useful for isolating "soft" crashes/unwanted behavior
like codegen blowing up.

Adds `--iree-flow-break-on-dispatch=` and `--iree-flow-trace-dispatch=`
flags with support for tracing/breaking on both dispatch symbols and
ordinals. Dispatch names are only partially matched against the symbols,
enabling things like `--iree-flow-trace-dispatch=conv_2d_nchw` to trace
specific kinds of compute ops.

Support for (nested) control flow and propagating return type changes
(in break mode) to function calls is still TODO as the need arises.

ScottTodd commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.h:
```diff
@@ -176,6 +176,10 @@ createOutlineDispatchRegionsPass();
 
+// Crops the dispatch calls at the given index (-1 for nothing)
+std::unique_ptr<InterfacePass<mlir::FunctionOpInterface>>
+createCropDispatchPipelinePass(int dispatchIndex = -1);
```
There was an RFC posted this week, and discussed at the MLIR open meeting yesterday, that may offer a more general way to implement this sort of compiler debugging: https://discourse.llvm.org/t/rfc-mlir-action-tracing-and-debugging-mlir-based-compilers/68679 ([slides/recording linked here](https://discourse.llvm.org/t/open-mlir-meeting-2-23-2023-mlir-actions-tracing-and-debugging-mlir-based-compilers/68680)). With that you could filter actions (pattern applications, passes, etc.) and choose to skip some - possibly even without needing to write C++ code or recompile the compiler.

qedawkins replied:
Oh cool, having these tools would probably eliminate the use case for this pass to debug compiler failures. Although I think having the ability to jump in and look at the output of a specific dispatch would be hard to do with this. Perhaps that's closer to something we want to be able to do in the runtime? So that the model doesn't need to be recompiled just to look at the output of a neighboring dispatch.

qedawkins committed:
Fix asan

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp:
```diff
@@ -34,7 +34,20 @@ static llvm::cl::opt<bool> clTraceDispatchTensors(
+    llvm::cl::desc("Enables tracing tensors at specified dispatches. Supports "
+                   "two modes; tracing the dispatch by ordinal before "
+                   "deduplication (@function_name:<index>) and tracing all "
+                   "occurances of the dispatch symbol."),
```
```suggestion
                   "occurrences of the dispatch symbol."),
```

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InsertDispatchDebugTargets.cpp:
```diff
@@ -0,0 +1,253 @@
+// Copyright 2020 The IREE Authors
```
```suggestion
// Copyright 2023 The IREE Authors
```

benvanik commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InsertDispatchDebugTargets.cpp:
```diff
@@ -0,0 +1,253 @@
+          entryPointName = (entryPointName + "::" + nestedRef.getValue()).str();
+        }
+        if (!traceDebugTarget.empty() &&
+            entryPointName.find(traceDebugTarget) != std::string::npos)
```
we'll want regex to allow for things like saying `dispatch_1` but not `dispatch_10` +`dispatch_11` too - llvm::Regex should make this fairly easy (make Regex at top of runOnOperation and then `match` here and below)

benvanik reviewed (changes requested):
This looks fantastic! Very sorry about the delay - if you'd like I can make the changes/rebase/land since it's been so long.

qedawkins committed:
rebase and address comments

Adds regex matching for symbol name
Also renames `iree-flow-break-on-dispatch` to `iree-flow-break-dispatch`
because it's much more convenient to just replace `trace` with `break`
when swapping modes
